ksqldb-functional-tests/src/test/resources/query-validation-tests/in.json

```diff
@@ -0,0 +1,223 @@
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, VAL INT) WITH (kafka_topic='input_topic', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT WHERE ID IN (123456789123456789,20,10,'not a number',false,12.0, ARRAY[1]);"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 10, "value": {}},
+        {"topic": "input_topic", "key": 12, "value": {}},
+        {"topic": "input_topic", "key": 123456789123456789, "value": {}},
+        {"topic": "input_topic", "key": 19, "value": {}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 10, "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": 123456789123456789, "value": {"VAL": null}}
+      ]
+    },
+    {
+      "name": "double",
+      "statements": [
+        "CREATE STREAM INPUT (ID DOUBLE KEY, VAL INT) WITH (kafka_topic='input_topic', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT WHERE ID IN (10, 10.1, '10.2', 10.30, '10.40', 'not a number', false, ARRAY[1]);"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 10.0, "value": {}},
+        {"topic": "input_topic", "key": 10.1, "value": {}},
+        {"topic": "input_topic", "key": 10.2, "value": {}},
+        {"topic": "input_topic", "key": 10.3, "value": {}},
+        {"topic": "input_topic", "key": 10.4, "value": {}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 10.1, "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": 10.2, "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": 10.4, "value": {"VAL": null}}
+      ]
+    },
+    {
+      "name": "string",
+      "statements": [
+        "CREATE STREAM INPUT (ID STRING KEY, VAL INT) WITH (kafka_topic='input_topic', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT WHERE ID IN ('k', 10, '10.1', 10.30, 'not a match', false, ARRAY[1]);"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": "k", "value": {}},
+        {"topic": "input_topic", "key": "no", "value": {}},
+        {"topic": "input_topic", "key": "10", "value": {}},
+        {"topic": "input_topic", "key": "10.0", "value": {}},
+        {"topic": "input_topic", "key": "10.1", "value": {}},
+        {"topic": "input_topic", "key": "10.10", "value": {}},
+        {"topic": "input_topic", "key": "false", "value": {}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": "k", "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": "10", "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": "10.1", "value": {"VAL": null}},
+        {"topic": "OUTPUT", "key": "false", "value": {"VAL": null}}
+      ]
+    },
+    {
+      "name": "decimal",
+      "statements": [
+        "CREATE STREAM INPUT (ID DECIMAL(4,2) KEY, VAL INT) WITH (kafka_topic='input_topic', format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT * FROM INPUT WHERE ID IN (10, 10.0, 10.10, 10.201, '10.30', '10.4', '10.400', 'not a match', false, ARRAY[1]);"
```

Comment:
Is it not considered a type error to mix types in the in list?  You would also think that it would be an error if the types weren't comparable.

Reply:
It's ... complicated.

A standard SQL engine might accept something like `ID IN ('10', 12, 11.9901, '76.887')` where `ID` is any numeric or string type, ass its coercing the types, but it might reject something like `ID IN ('10', 12, 11.9901, 'not a number')` if `ID` is a numeric type, but not if its a string type.

Likewise, a standard SQL engine might convert `'TRUE'`, `'true'`, `'t'` to `true`, etc.  

So it should be possible to do `textCol IN (10, true, 'hello')`, as `textCol` could contain `'10'`, `'true'` or `'hello'`.

And of course, the list may contain expressions as well we literals. If `ID` is a numeric type, and one of the expressions is a `STRING`, then the content of that string may or may not be a number.  So while I could of added code to detect string literals that can't be converted to the required type, and fail, given we can't do the same for non-literal expressions... I took the easy way out! I was just trying to 'rattle an initial implementation off'.

We could also look at the expression type and all the types in the `IN` element list and fail if they can never be compatible. e.g. `ID IN (true, false)` where `ID` is neither a string or a boolean.  But again, this is more work, though maybe better UX.

What say you??

Reply:
It is a bit complicated.

I agree that special-casing literals to catch certain type error cases that you could not catch up front with with general expressions doesn't seem like a high value change to make. That's why relying on strict types would be nice because you know that at analysis time and are ensured that whatever evaluated expressions you see during execution must conform to expected types.  But, if a standard SQL engine does a lot of this coercion, then I agree people may expect it.

I'd always lean towards being stricter and avoiding coercions, but if that's not an option, then what you have seems like a reasonable tradeoff.